### ISSUE TITLE
AutoModelParallel: automatic multi-GPU/TPU sharding for any Backbone (JAX)

### ISSUE THREAD

**Issue by yassienashrafwasfy** [keras-team/keras-hub#2806]
I am looking for a way that make inference & training easier in keras-Hub without writing a lot of complex regex in layout map
## What

    backbone = keras_hub.models.Backbone.from_preset("mistral_7b_en", sharding="auto")

One argument shards any backbone across all available devices.

Verified on real hardware, two accelerator families, same code path:

- Kaggle 2x T4: float16, weight bytes split 50/50 across the GPUs. The
  model does not fit on a single 16 GB T4.
- Kaggle TPU v5e-8: float32, 8-way sharding, ~29 GB of weights across
  8x 16 GB chips. Does not fit on a single chip either.

## Why

Only 5 backbones support model parallelism today, each via a hand-written
get_layout_map of regexes over variable paths. A rename silently breaks the
match and the model loads fully replicated (OOM instead of an error). HF has
the same ceiling: tp_plan="auto" reads hand-written per-model dicts, ~15
architectures covered, the rest get nothing.

## How

Layouts are derived from variable shapes, not names: replicate rank < 2 and
tiny variables, otherwise shard the largest axis divisible by the device
count, ties to the trailing axis. On rectangular kernels this reproduces the
Megatron colwise/rowwise pairing automatically. GSPMD guarantees any valid
layout gives identical numerics (tested sharded vs unsharded), so the worst
case is extra communication on square/fused kernels, never wrong results,
and a LayoutMap override recovers a hand-tuned layout per variable.
Optimizer states share their weight's shape so they inherit the same layout
for free. JAX only, clear error elsewhere.

## Tests

    KERAS_BACKEND=jax JAX_PLATFORMS=cpu \
    XLA_FLAGS=--xla_force_host_platform_device_count=4 \
    python -m pytest keras_hub/src/distribution/auto_model_parallel_test.py

Shape rule units, BertBackbone layout assertions, override precedence,
numerics parity, from_preset round trip, error paths. Plus the two real
hardware runs above.

## Files changed

- keras_hub/src/distribution/auto_model_parallel.py: the feature
- keras_hub/src/distribution/auto_model_parallel_test.py: tests
- keras_hub/src/models/backbone.py: sharding argument on from_preset
- keras_hub/api/: generated exports via api_gen.py

### TPUv5e-8
<img width="610" height="225" alt="Image" src="https://github.com/user-attachments/assets/9a1e99f5-a25f-47ed-9d4e-5b3a8f381102" />

### 2xT4 Nvidia Cards
<img width="538" height="77" alt="Image" src="https://github.com/user-attachments/assets/f6ec6567-752a-413e-a604-a9a6d9380bb7" />

**Comment by yassienashrafwasfy** [comment]
[Commit Mention](https://github.com/keras-team/keras-hub/commit/a1915295e5c29cd0e3e9e18dd322c637c9913f4c)

**Comment by divyashreepathihalli** [comment]
The team is currently looking into autosharding features. Stay tuned for updates
